apis/apps/v1alpha1/workloadspread_types.go

```diff
@@ -0,0 +1,210 @@
+)
+
+// WorkloadSpreadScheduleStrategy defines the schedule performance of WorkloadSpread
+type WorkloadSpreadScheduleStrategy struct {
+	// Type indicates the type of the WorkloadSpreadScheduleStrategy.
+	// Default is Fixed
+	// +optional
+	Type WorkloadSpreadScheduleStrategyType `json:"type,omitempty"`
+
+	// Adaptive is used to communicate parameters when Type is AdaptiveWorkloadSpreadScheduleStrategyType.
+	// +optional
+	Adaptive *AdaptiveWorkloadSpreadStrategy `json:"adaptive,omitempty"`
+}
+
+// AdaptiveWorkloadSpreadStrategy is used to communicate parameters when Type is AdaptiveWorkloadSpreadScheduleStrategyType.
+type AdaptiveWorkloadSpreadStrategy struct {
+	// DisableSimulationSchedule indicates whether to disable the feature of simulation schedule.
+	// Default is false.
+	// Webhook can take a simple general predicates to check whether Pod can be scheduled into this subset,
+	// but it just considers the Node resource and cannot replace scheduler to do richer predicates practically.
+	// +optional
+	DisableSimulationSchedule bool `json:"disableSimulationSchedule,omitempty"`
+
+	// RescheduleCriticalSeconds indicates how long controller will reschedule a schedule failed Pod to the subset that has
+	// redundant capacity after the subset where the Pod lives. If a Pod was scheduled failed and still in a pending status
+	// over RescheduleCriticalSeconds duration, the controller will reschedule it to the suitable subset.
+	// +optional
+	RescheduleCriticalSeconds *int32 `json:"rescheduleCriticalSeconds,omitempty"`
+}
+
+// WorkloadSpreadSubset defines the details of a subset.
+type WorkloadSpreadSubset struct {
+	// Name should be unique between all of the subsets under one WorkloadSpread.
+	Name string `json:"name"`
+
+	// Indicates the node required selector to form the subset.
+	// +optional
+	RequiredNodeSelectorTerm *corev1.NodeSelectorTerm `json:"requiredNodeSelectorTerm,omitempty"`
+
+	// Indicates the node preferred selector to form the subset.
+	// +optional
+	PreferredNodeSelectorTerms []corev1.PreferredSchedulingTerm `json:"preferredNodeSelectorTerms,omitempty"`
+
+	// Indicates the tolerations the pods under this subset have.
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
+	// MaxReplicas indicates the desired max replicas of this subset.
+	// +optional
+	MaxReplicas *intstr.IntOrString `json:"maxReplicas,omitempty"`
+
+	// Patch indicates the patch metadata to the Pod.
+	// +optional
+	Patch runtime.RawExtension `json:"patch,omitempty"`
+}
+
+// WorkloadSpreadStatus defines the observed state of WorkloadSpread.
+type WorkloadSpreadStatus struct {
+	// ObservedGeneration is the most recent generation observed for this WorkloadSpread. It corresponds to the
+	// WorkloadSpread's generation, which is updated on mutation by the API Server.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
```

Comment:
Is Generation equal to ResourceVersion in object meta?

Reply:
Generation equals to generation in object meta